File: ts/src/ascii.ts
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

/*  ascii.ts
 *  Vertical-flow ASCII renderer for railroad diagrams. Same flow model as
 *  the SVG renderer (sequences stack downward, choices fan sideways, loops
 *  return on the right), painted onto a character grid.
 *
 *  Rails are tracked as direction bits per cell (Up/Down/Left/Right) so
 *  junctions resolve automatically to the right box-drawing glyph. Boxes
 *  draw literal corners/sides. `opts.ascii` (or the CLI `--ascii-plain`)
 *  swaps the Unicode glyph set for plain `| - +`.
 */

import { RailroadNode, GrammarModel, Item, norm, RailroadError } from './model'


export type AsciiOptions = {
  ascii?: boolean   // true => pure-ASCII glyphs (| - +) instead of Unicode
}

const U = 1, D = 2, L = 4, R = 8

const GLYPH: { [bits: number]: string } = {
  3: '│', 12: '─', 10: '┌', 6: '┐', 9: '└', 5: '┘',
  11: '├', 7: '┤', 14: '┬', 13: '┴', 15: '┼',
  1: '│', 2: '│', 4: '─', 8: '─',
}
function glyphFor(bits: number, plain: boolean): string {
  if (0 === bits) return ' '
  if (plain) {
    const v = bits & 3, h = bits & 12
    if (v && h) return '+'
    return v ? '|' : '-'
  }
  return GLYPH[bits] || ' '
}

function glyphs(plain: boolean) {
  return plain
    ? { tl: '+', tr: '+', bl: '+', br: '+', rtl: '+', rtr: '+', rbl: '+', rbr: '+', v: '|' }
    : { tl: '┌', tr: '┐', bl: '└', br: '┘', rtl: '╭', rtr: '╮', rbl: '╰', rbr: '╯', v: '│' }
}


class Canvas {
  private bits: number[][] = []
  private lit: (string | null)[][] = []
  private grow(r: number, c: number) {
    while (this.bits.length <= r) { this.bits.push([]); this.lit.push([]) }
    const row = this.bits[r], lrow = this.lit[r]
    while (row.length <= c) { row.push(0); lrow.push(null) }
  }
  line(r: number, c: number, mask: number) { this.grow(r, c); this.bits[r][c] |= mask }
  put(r: number, c: number, ch: string) { this.grow(r, c); this.lit[r][c] = ch }
  text(r: number, c: number, s: string) { for (let i = 0; i < s.length; i++) this.put(r, c + i, s[i]) }
  vline(r1: number, r2: number, c: number) {
    if (r1 > r2) { const t = r1; r1 = r2; r2 = t }
    for (let r = r1; r <= r2; r++) this.line(r, c, (r > r1 ? U : 0) | (r < r2 ? D : 0))
  }
  hline(c1: number, c2: number, r: number) {
    if (c1 > c2) { const t = c1; c1 = c2; c2 = t }
    for (let c = c1; c <= c2; c++) this.line(r, c, (c > c1 ? L : 0) | (c < c2 ? R : 0))
  }
  render(plain: boolean): string {
    return this.bits.map((row, r) => {
      let line = ''
      for (let c = 0; c < row.length; c++) {
        const lc = this.lit[r][c]
        line += null != lc ? lc : glyphFor(row[c], plain)
      }
      return line.replace(/\s+$/, '')
    }).join('\n')
  }
}


// ---- measure model -------------------------------------------------
type M = {
  cols: number
  rows: number
  entryCol: number
  exitCol: number
  paint: (cv: Canvas, x: number, y: number) => void
}

const VG = 1   // vertical gap rows between stacked items
const HG = 3   // horizontal gap cols between choice branches


function boxM(text: string, terminal: boolean, G: ReturnType<typeof glyphs>): M {
  const inner = ' ' + text + ' '
  const w = inner.length + 2
  const mid = Math.floor(w / 2)
  return {
    cols: w, rows: 3, entryCol: mid, exitCol: mid,
    paint(cv, x, y) {
      cv.put(y, x, terminal ? G.rtl : G.tl)
      cv.put(y, x + w - 1, terminal ? G.rtr : G.tr)
      cv.hline(x + 1, x + w - 2, y)
      cv.put(y + 1, x, G.v)
      cv.put(y + 1, x + w - 1, G.v)
      cv.text(y + 1, x + 1, inner)
      cv.put(y + 2, x, terminal ? G.rbl : G.bl)
      cv.put(y + 2, x + w - 1, terminal ? G.rbr : G.br)
      cv.hline(x + 1, x + w - 2, y + 2)
    },
  }
}

function commentM(text: string): M {
  const s = '/* ' + text + ' */'
  return {
    cols: s.length, rows: 1, entryCol: Math.floor(s.length / 2), exitCol: Math.floor(s.length / 2),
    paint(cv, x, y) { cv.text(y, x, s) },
  }
}

function skipM(): M {
  return { cols: 1, rows: 1, entryCol: 0, exitCol: 0, paint(cv, x, y) { cv.line(y, x, U | D) } }
}

function seqM(children: M[]): M {
  if (0 === children.length) return skipM()
  if (1 === children.length) return children[0]
  const railCol = Math.max(...children.map((c) => c.entryCol))
  const offs = children.map((c) => railCol - c.entryCol)
  const cols = children.reduce((m, c, i) => Math.max(m, offs[i] + c.cols), 0)
  const rows = children.reduce((a, c) => a + c.rows, 0) + VG * (children.length - 1)
  return {
    cols, rows, entryCol: railCol, exitCol: railCol,
    paint(cv, x, y) {
      let cy = y
      children.forEach((c, i) => {
        if (i > 0) { cv.vline(cy - VG - 1, cy, x + railCol) }
        c.paint(cv, x + offs[i], cy)
        cy += c.rows + VG
      })
    },
  }
}

function choiceM(branches: M[]): M {
  if (1 === branches.length) return branches[0]
  const n = branches.length
  const cols = branches.reduce((a, c) => a + c.cols, 0) + HG * (n - 1)
  const maxR = Math.max(...branches.map((c) => c.rows))
  const rows = 1 + maxR + 1
  const bxs: number[] = []
  let bx = 0
  branches.forEach((c) => { bxs.push(bx); bx += c.cols + HG })
  const relCenters = branches.map((c, i) => bxs[i] + c.entryCol)
  const entryCol = Math.round((relCenters[0] + relCenters[n - 1]) / 2)
  return {
    cols, rows, entryCol, exitCol: entryCol,
    paint(cv, x, y) {
      const splitRow = y
      const branchTop = y + 1
      const mergeRow = y + 1 + maxR
      const centers = relCenters.map((c) => x + c)
      const lo = Math.min(x + entryCol, ...centers)
      const hi = Math.max(x + entryCol, ...centers)
      cv.hline(lo, hi, splitRow)
      cv.hline(lo, hi, mergeRow)
      branches.forEach((c, i) => {
        const cx = x + bxs[i]
        const ce = cx + c.entryCol
        cv.vline(splitRow, branchTop, ce)
        c.paint(cv, cx, branchTop)
        cv.vline(branchTop + c.rows - 1, mergeRow, cx + c.exitCol)
      })
    },
  }
}

function oneOrMoreM(item: M, repLabel: string): M {
  const railGap = repLabel ? repLabel.length + 1 : 0
  const cols = item.cols + 1 + railGap
  const rows = item.rows + 2
  const railCol = item.cols   // one column right of the item block
  return {
    cols, rows, entryCol: item.entryCol, exitCol: item.exitCol,
    paint(cv, x, y) {
      const topRow = y
      const itemTop = y + 1
      const itemBot = y + item.rows           // last item row
      const botRow = y + item.rows + 1
      item.paint(cv, x, itemTop)
      cv.vline(topRow, itemTop, x + item.entryCol)
      cv.vline(itemBot, botRow, x + item.exitCol)
      const rc = x + railCol
      cv.hline(x + item.entryCol, rc, topRow)
      cv.hline(x + item.exitCol, rc, botRow)
      cv.vline(topRow, botRow, rc)
      if (repLabel) cv.text(y + Math.floor(rows / 2), rc + 1, repLabel)
    },
  }
}

function labelOf(node?: RailroadNode): string {
  if (!node) return ''
  if ('terminal' === node.kind || 'nonterminal' === node.kind) return node.text
  return '+'
}

function measure(node: RailroadNode, G: ReturnType<typeof glyphs>): M {
  switch (node.kind) {
    case 'terminal': return boxM(JSON.stringify(node.text), true, G)
    case 'nonterminal': return boxM(node.text, false, G)
    case 'comment': return commentM(node.text)
    case 'skip': return skipM()
    case 'seq': return seqM(node.items.map((n) => measure(n, G)))
    case 'choice': return choiceM(node.items.map((n) => measure(n, G)))
    case 'optional': return choiceM([measure(node.item, G), skipM()])
    case 'oneOrMore': return oneOrMoreM(measure(node.item, G), labelOf(node.rep))
    case 'zeroOrMore': return choiceM([oneOrMoreM(measure(node.item, G), labelOf(node.rep)), skipM()])
    case 'diagram': return seqM(node.items.map((n) => measure(n, G)))
    default:
      throw new RailroadError(
        'railroad: unknown node kind ' + JSON.stringify((node as any).kind), node)
  }
}


// Render a single node to an ASCII block.
export function renderNodeAscii(node: Item, opts: AsciiOptions = {}): string {
  const plain = !!opts.ascii
  const G = glyphs(plain)
  const cv = new Canvas()
  const m = measure(norm(node), G)
  // top + bottom rail caps, joined into the node's entry/exit cells
  cv.line(0, m.entryCol, D)
  m.paint(cv, 0, 1)
  cv.line(1, m.entryCol, U)
  cv.line(m.rows, m.exitCol, D)
  cv.line(m.rows + 1, m.exitCol, U)
  return cv.render(plain)
}


// Render a whole grammar: each rule as a titled vertical block.
export function modelToAscii(model: GrammarModel, opts: AsciiOptions = {}): string {
  const plain = !!opts.ascii
  const G = glyphs(plain)
  const names = orderRules(model)
  const blocks: string[] = []
  for (const name of names) {
    const cv = new Canvas()
    const m = measure(model.rules[name], G)
    cv.line(0, m.entryCol, D)
    m.paint(cv, 0, 1)
    cv.line(1, m.entryCol, U)
    cv.line(m.rows, m.exitCol, D)
    cv.line(m.rows + 1, m.exitCol, U)
    const diagram = cv.render(plain)
    blocks.push(name + ':\n' + diagram)
  }
  const keyBlock = (title: string, entries: { token: string; meaning: string }[]) => {
    const w = Math.max(...entries.map((e) => e.token.length))
    return title + ':\n' +
      entries.map((e) => '  ' + e.token.padEnd(w) + ' = ' + e.meaning).join('\n')
  }
  if (model.legend && model.legend.length) blocks.push(keyBlock('Tokens', model.legend))
  // Tokens the lexer silently skips (IGNORE set) — never appear in a rule.
  if (model.ignored && model.ignored.length) blocks.push(keyBlock('Ignored tokens', model.ignored))
  return blocks.join('\n\n')
}

function orderRules(model: GrammarModel): string[] {
  const names = Object.keys(model.rules)
  if (model.start && names.includes(model.start)) {
    return [model.start, ...names.filter((n) => n !== model.start)]
  }
  return names
}

File: rs/src/ascii.rs
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

//! Vertical-flow ASCII renderer for railroad diagrams. Same flow model as
//! the SVG renderer (sequences stack downward, choices fan sideways,
//! loops return on the right), painted onto a character grid.
//!
//! Rails are tracked as direction bits per cell (up, down, left, right)
//! so junctions resolve automatically to the right box-drawing glyph.
//! Boxes draw literal corners and sides. [`AsciiOptions::ascii`] (the
//! CLI `--ascii-plain`) swaps the Unicode glyph set for plain `| - +`.
//!
//! This file mirrors `ts/src/ascii.ts`.

use crate::model::{json_string, GrammarModel, LegendEntry, RailroadNode};

/// Options for the ASCII renderer.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct AsciiOptions {
    /// Pure-ASCII glyphs (`| - +`) instead of Unicode box drawing. The
    /// TypeScript spelling is `{ ascii: true }`; the Go field is `Plain`.
    pub ascii: bool,
}

impl AsciiOptions {
    /// The plain `| - +` glyph set.
    pub fn plain() -> Self {
        AsciiOptions { ascii: true }
    }
}

const U: u8 = 1;
const D: u8 = 2;
const L: u8 = 4;
const R: u8 = 8;

fn glyph_for(bits: u8, plain: bool) -> char {
    if bits == 0 {
        return ' ';
    }
    if plain {
        let vertical = bits & 3 != 0;
        let horizontal = bits & 12 != 0;
        return match (vertical, horizontal) {
            (true, true) => '+',
            (true, false) => '|',
            _ => '-',
        };
    }
    match bits {
        1..=3 => '│',
        12 | 4 | 8 => '─',
        10 => '┌',
        6 => '┐',
        9 => '└',
        5 => '┘',
        11 => '├',
        7 => '┤',
        14 => '┬',
        13 => '┴',
        15 => '┼',
        _ => ' ',
    }
}

#[derive(Clone, Copy)]
struct Glyphs {
    tl: char,
    tr: char,
    bl: char,
    br: char,
    rtl: char,
    rtr: char,
    rbl: char,
    rbr: char,
    v: char,
}

fn glyphs(plain: bool) -> Glyphs {
    if plain {
        Glyphs {
            tl: '+',
            tr: '+',
            bl: '+',
            br: '+',
            rtl: '+',
            rtr: '+',
            rbl: '+',
            rbr: '+',
            v: '|',
        }
    } else {
        Glyphs {
            tl: '┌',
            tr: '┐',
            bl: '└',
            br: '┘',
            rtl: '╭',
            rtr: '╮',
            rbl: '╰',
            rbr: '╯',
            v: '│',
        }
    }
}

/// A literal cell: a character, or the trailing half of a character
/// that JavaScript counts as two UTF-16 units. The character is printed
/// by the cell before it; the trailing half prints nothing, so the row
/// renders to the same string the two-cell JavaScript canvas gives.
#[derive(Clone, Copy)]
enum Lit {
    Char(char),
    Trail,
}

/// The character grid: per-cell direction bits resolve to junction
/// glyphs, and a literal cell overrides them.
#[derive(Default)]
struct Canvas {
    bits: Vec<Vec<u8>>,
    lit: Vec<Vec<Option<Lit>>>,
}

impl Canvas {
    fn grow(&mut self, r: usize, c: usize) {
        while self.bits.len() <= r {
            self.bits.push(Vec::new());
            self.lit.push(Vec::new());
        }
        let row = &mut self.bits[r];
        let lrow = &mut self.lit[r];
        while row.len() <= c {
            row.push(0);
            lrow.push(None);
        }
    }

    fn line(&mut self, r: usize, c: usize, mask: u8) {
        self.grow(r, c);
        self.bits[r][c] |= mask;
    }

    fn put(&mut self, r: usize, c: usize, ch: char) {
        self.grow(r, c);
        self.lit[r][c] = Some(Lit::Char(ch));
    }

    /// Write `s` from column `c`, one cell per UTF-16 unit: a character
    /// outside the Basic Multilingual Plane takes two cells, as it does
    /// in TypeScript, so every width and rail column agrees.
    fn text(&mut self, r: usize, c: usize, s: &str) {
        let mut col = c;
        for ch in s.chars() {
            self.put(r, col, ch);
            col += 1;
            if ch.len_utf16() == 2 {
                self.grow(r, col);
                self.lit[r][col] = Some(Lit::Trail);
                col += 1;
            }
        }
    }

    fn vline(&mut self, r1: usize, r2: usize, c: usize) {
        let (r1, r2) = if r1 > r2 { (r2, r1) } else { (r1, r2) };
        for r in r1..=r2 {
            let mask = if r > r1 { U } else { 0 } | if r < r2 { D } else { 0 };
            self.line(r, c, mask);
        }
    }

    fn hline(&mut self, c1: usize, c2: usize, r: usize) {
        let (c1, c2) = if c1 > c2 { (c2, c1) } else { (c1, c2) };
        for c in c1..=c2 {
            let mask = if c > c1 { L } else { 0 } | if c < c2 { R } else { 0 };
            self.line(r, c, mask);
        }
    }

    fn render(&self, plain: bool) -> String {
        self.bits
            .iter()
            .enumerate()
            .map(|(r, row)| {
                let mut line = String::with_capacity(row.len());
                for (c, bits) in row.iter().enumerate() {
                    match self.lit[r][c] {
                        Some(Lit::Char(ch)) => line.push(ch),
                        Some(Lit::Trail) => {}
                        None => line.push(glyph_for(*bits, plain)),
                    }
                }
                js_trim_end(&line).to_string()
            })
            .collect::<Vec<String>>()
            .join("\n")
    }
}

/// Whether `ch` is whitespace to a JavaScript `\s`: the ECMAScript
/// WhiteSpace and LineTerminator sets. This is not Rust's `White_Space`
/// (which has NEXT LINE and lacks the byte order mark), and the renderer
/// trims a row's end with this so a rail label ending in such a character
/// trims exactly as it does in TypeScript.
fn is_js_whitespace(ch: char) -> bool {
    matches!(
        ch,
        '\t' | '\n' | '\u{0B}' | '\u{0C}' | '\r' | ' ' | '\u{A0}' | '\u{1680}' | '\u{2000}'
            ..='\u{200A}'
                | '\u{2028}'
                | '\u{2029}'
                | '\u{202F}'
                | '\u{205F}'
                | '\u{3000}'
                | '\u{FEFF}'
    )
}

/// `line` without its trailing JavaScript whitespace: the
/// `replace(/\s+$/, '')` of the TypeScript canvas.
fn js_trim_end(line: &str) -> &str {
    line.trim_end_matches(is_js_whitespace)
}

// ---- measure model -------------------------------------------------

type Paint = Box<dyn Fn(&mut Canvas, usize, usize)>;

/// A measured node: its size, where the rail enters and leaves, and how
/// to paint it at a position.
struct Measure {
    cols: usize,
    rows: usize,
    entry_col: usize,
    exit_col: usize,
    paint: Paint,
}

/// Vertical gap rows between stacked items.
const VG: usize = 1;
/// Horizontal gap cols between choice branches.
const HG: usize = 3;

/// The width of `s` in cells: its length in UTF-16 units, which is what
/// the TypeScript `String.length` measures, so a label outside the Basic
/// Multilingual Plane is two cells wide in both.
fn width(s: &str) -> usize {
    s.encode_utf16().count()
}

fn box_m(text: &str, is_terminal: bool, g: Glyphs) -> Measure {
    let inner = format!(" {text} ");
    let w = width(&inner) + 2;
    let mid = w / 2;
    Measure {
        cols: w,
        rows: 3,
        entry_col: mid,
        exit_col: mid,
        paint: Box::new(move |cv, x, y| {
            let (tl, tr, bl, br) = if is_terminal {
                (g.rtl, g.rtr, g.rbl, g.rbr)
            } else {
                (g.tl, g.tr, g.bl, g.br)
            };
            cv.put(y, x, tl);
            cv.put(y, x + w - 1, tr);
            cv.hline(x + 1, x + w - 2, y);
            cv.put(y + 1, x, g.v);
            cv.put(y + 1, x + w - 1, g.v);
            cv.text(y + 1, x + 1, &inner);
            cv.put(y + 2, x, bl);
            cv.put(y + 2, x + w - 1, br);
            cv.hline(x + 1, x + w - 2, y + 2);
        }),
    }
}

fn comment_m(text: &str) -> Measure {
    let s = format!("/* {text} */");
    let w = width(&s);
    Measure {
        cols: w,
        rows: 1,
        entry_col: w / 2,
        exit_col: w / 2,
        paint: Box::new(move |cv, x, y| cv.text(y, x, &s)),
    }
}

fn skip_m() -> Measure {
    Measure {
        cols: 1,
        rows: 1,
        entry_col: 0,
        exit_col: 0,
        paint: Box::new(|cv, x, y| cv.line(y, x, U | D)),
    }
}

fn seq_m(mut children: Vec<Measure>) -> Measure {
    if children.is_empty() {
        return skip_m();
    }
    if children.len() == 1 {
        return children.remove(0);
    }
    let rail_col = children.iter().map(|c| c.entry_col).max().unwrap_or(0);
    let offs: Vec<usize> = children.iter().map(|c| rail_col - c.entry_col).collect();
    let cols = children
        .iter()
        .zip(&offs)
        .map(|(c, off)| off + c.cols)
        .max()
        .unwrap_or(0);
    let rows = children.iter().map(|c| c.rows).sum::<usize>() + VG * (children.len() - 1);
    Measure {
        cols,
        rows,
        entry_col: rail_col,
        exit_col: rail_col,
        paint: Box::new(move |cv, x, y| {
            let mut cy = y;
            for (i, c) in children.iter().enumerate() {
                if i > 0 {
                    cv.vline(cy - VG - 1, cy, x + rail_col);
                }
                (c.paint)(cv, x + offs[i], cy);
                cy += c.rows + VG;
            }
        }),
    }
}

fn choice_m(mut branches: Vec<Measure>) -> Measure {
    // A choice with no branches cannot come from a constructor or from
    // JSON, but the enum can be built by hand; it renders as a bypass
    // rather than indexing a branch that is not there.
    if branches.is_empty() {
        return skip_m();
    }
    if branches.len() == 1 {
        return branches.remove(0);
    }
    let n = branches.len();
    let cols = branches.iter().map(|c| c.cols).sum::<usize>() + HG * (n - 1);
    let max_r = branches.iter().map(|c| c.rows).max().unwrap_or(0);
    let rows = 1 + max_r + 1;
    let mut bxs = Vec::with_capacity(n);
    let mut bx = 0;
    for c in &branches {
        bxs.push(bx);
        bx += c.cols + HG;
    }
    let rel_centers: Vec<usize> = branches
        .iter()
        .zip(&bxs)
        .map(|(c, bx)| bx + c.entry_col)
        .collect();
    // `Math.round` of a half rounds up, which for these non-negative
    // integers is the ceiling of the midpoint.
    let entry_col = (rel_centers[0] + rel_centers[n - 1]).div_ceil(2);
    Measure {
        cols,
        rows,
        entry_col,
        exit_col: entry_col,
        paint: Box::new(move |cv, x, y| {
            let split_row = y;
            let branch_top = y + 1;
            let merge_row = y + 1 + max_r;
            let centers: Vec<usize> = rel_centers.iter().map(|c| x + c).collect();
            let lo = centers.iter().copied().fold(x + entry_col, usize::min);
            let hi = centers.iter().copied().fold(x + entry_col, usize::max);
            cv.hline(lo, hi, split_row);
            cv.hline(lo, hi, merge_row);
            for (i, c) in branches.iter().enumerate() {
                let cx = x + bxs[i];
                let ce = cx + c.entry_col;
                cv.vline(split_row, branch_top, ce);
                (c.paint)(cv, cx, branch_top);
                cv.vline(branch_top + c.rows - 1, merge_row, cx + c.exit_col);
            }
        }),
    }
}

fn one_or_more_m(item: Measure, rep_label: String) -> Measure {
    let rail_gap = if rep_label.is_empty() {
        0
    } else {
        width(&rep_label) + 1
    };
    let cols = item.cols + 1 + rail_gap;
    let rows = item.rows + 2;
    // One column right of the item block.
    let rail_col = item.cols;
    Measure {
        cols,
        rows,
        entry_col: item.entry_col,
        exit_col: item.exit_col,
        paint: Box::new(move |cv, x, y| {
            let top_row = y;
            let item_top = y + 1;
            // The last item row.
            let item_bot = y + item.rows;
            let bot_row = y + item.rows + 1;
            (item.paint)(cv, x, item_top);
            cv.vline(top_row, item_top, x + item.entry_col);
            cv.vline(item_bot, bot_row, x + item.exit_col);
            let rc = x + rail_col;
            cv.hline(x + item.entry_col, rc, top_row);
            cv.hline(x + item.exit_col, rc, bot_row);
            cv.vline(top_row, bot_row, rc);
            if !rep_label.is_empty() {
                cv.text(y + rows / 2, rc + 1, &rep_label);
            }
        }),
    }
}

fn label_of(node: Option<&RailroadNode>) -> String {
    match node {
        None => String::new(),
        Some(RailroadNode::Terminal { text }) | Some(RailroadNode::NonTerminal { text }) => {
            text.clone()
        }
        Some(_) => "+".to_string(),
    }
}

fn measure(node: &RailroadNode, g: Glyphs) -> Measure {
    match node {
        RailroadNode::Terminal { text } => box_m(&json_string(text), true, g),
        RailroadNode::NonTerminal { text } => box_m(text, false, g),
        RailroadNode::Comment { text } => comment_m(text),
        RailroadNode::Skip => skip_m(),
        RailroadNode::Seq { items } | RailroadNode::Diagram { items } => {
            seq_m(items.iter().map(|n| measure(n, g)).collect())
        }
        RailroadNode::Choice { items } => choice_m(items.iter().map(|n| measure(n, g)).collect()),
        RailroadNode::Optional { item } => choice_m(vec![measure(item, g), skip_m()]),
        RailroadNode::OneOrMore { item, rep } => {
            one_or_more_m(measure(item, g), label_of(rep.as_deref()))
        }
        RailroadNode::ZeroOrMore { item, rep } => choice_m(vec![
            one_or_more_m(measure(item, g), label_of(rep.as_deref())),
            skip_m(),
        ]),
    }
}

/// Paint a measured node with its top and bottom rail caps, joined into
/// the node's entry and exit cells.
fn paint_block(m: &Measure, plain: bool) -> String {
    let mut cv = Canvas::default();
    cv.line(0, m.entry_col, D);
    (m.paint)(&mut cv, 0, 1);
    cv.line(1, m.entry_col, U);
    cv.line(m.rows, m.exit_col, D);
    cv.line(m.rows + 1, m.exit_col, U);
    cv.render(plain)
}

/// Render a single node to an ASCII block.
pub fn render_node_ascii(node: &RailroadNode, opts: &AsciiOptions) -> String {
    let plain = opts.ascii;
    let m = measure(node, glyphs(plain));
    paint_block(&m, plain)
}

/// Render a whole grammar: each rule as a titled vertical block, then
/// the token key and the ignored-token key when the model carries them.
pub fn model_to_ascii(model: &GrammarModel, opts: &AsciiOptions) -> String {
    let plain = opts.ascii;
    let g = glyphs(plain);
    let mut blocks: Vec<String> = Vec::new();
    for name in model.rule_order() {
        let m = measure(&model.rules[name], g);
        blocks.push(format!("{name}:\n{}", paint_block(&m, plain)));
    }
    if !model.legend.is_empty() {
        blocks.push(key_block("Tokens", &model.legend));
    }
    // Tokens the lexer silently skips (the IGNORE set) never appear in a
    // rule, so they get their own key.
    if !model.ignored.is_empty() {
        blocks.push(key_block("Ignored tokens", &model.ignored));
    }
    blocks.join("\n\n")
}

fn key_block(title: &str, entries: &[LegendEntry]) -> String {
    let w = entries.iter().map(|e| width(&e.token)).max().unwrap_or(0);
    let lines: Vec<String> = entries
        .iter()
        .map(|e| {
            let pad = " ".repeat(w - width(&e.token));
            format!("  {}{pad} = {}", e.token, e.meaning)
        })
        .collect();
    format!("{title}:\n{}", lines.join("\n"))
}
